Display admin email in error message for password reminder

petermaksymo opened the pull request:
To solve #626 

Let me know if this is needed for other error messages of if there is a cleaner way to implement it. For example, not sure if it is needed for [this message](https://github.com/spiral-project/ihatemoney/blob/master/ihatemoney/web.py#L337-L343).

almet commented on ihatemoney/web.py:
```diff
@@ -363,12 +363,24 @@ def remind_password():
                         "with password reset instructions. "
                         "Please check the email configuration of the server "
-                        "or contact the administrator."
+                        "or contact the administrator. {}".format(error_extension)
```
It seems to me that we're repeating ourselves here. The final sentence should be something like "Sorry, […] or contact the administrator at `email`".

almet reviewed (changes requested):
I added a few comments that needs to be addressed before we can merge. It makes me think that we should probably create a helper for these messages, and display the admin email in multiple places, for the different error messages. I feel this would help to tidy a bit the codebase : we currently have the same try-sending-email-or-display-error logic in multiple places.

Also, it makes me wonder if we should add a note in the documentation about this admin email setting, saying that it will be a public email used to receive help support.

petermaksymo commented on ihatemoney/web.py:
```diff
@@ -363,12 +363,24 @@ def remind_password():
+                error_extension = ""
+                if (
+                    admin_email
+                    and admin_email[1] != "budget@notmyidea.org"
```
This makes sense. This sounds like I should change `MAIL_DEFAULT_SENDER` throughout the repository to a generic admin@email.com or something along those lines?

almet replied:
Yes that seems good !

almet replied:
I can't find your pull request, but I thought you did, right? I've read about using `example.com` rather than other domain names today, see the rationale here : https://news.ycombinator.com/item?id=29623646

petermaksymo replied:
Thanks for bringing this up, I didn't know there was a difference. Changes made.

petermaksymo commented on ihatemoney/web.py:
```diff
@@ -363,12 +363,24 @@ def remind_password():
+                if (
+                    admin_email
+                    and admin_email[1] != "budget@notmyidea.org"
+                    and current_app.config.get("ALLOW_PUBLIC_PROJECT_CREATION")
```
This check is for only adding the email address to the warning message, not for sending the whole message. From my understanding, this was part of the issue #626:

> don't do that if public project creation is forbidden (because it would reveal more information to unauthentified users)

Maybe I misunderstood or am using the wrong variable?

almet replied:
Ha okay, I understand better. I feel this is more a side effect than the correct usage of this variable. We probably should add another configuration variable for this, no ?

petermaksymo commented:
Thanks for the feedback! I'll probably work on these changes and update the PR on Wednesday.

petermaksymo commented:
I've changed the default admin email as discussed above and refactored the error message logic into a helper function. I'm not too sure on what to name a new environment variable to replace the use of `ALLOW_PUBLIC_PROJECT_CREATION` in the helper.

almet commented:
Maybe something like `SHOW_ADMIN_EMAIL`? Don't forget to document it 👍

almet commented on docs/configuration.md:
```diff
@@ -82,6 +82,15 @@ emails.
+## SHOW_ADMIN_EMAIL
+
+A boolean that determines whether the admin email (`MAIL_DEFAULT_SENDER`) is
+shown in error messages.
```
You could also add the fact that if `MAIL_DEFAULT_SENDER` is not set, no email is shown.

almet commented:
Thanks, that works for me. I'm not sure about the fact that this means that all deployed instance will get their admin email shown to the public by default without a notice to them.

I also feel that adding this in the version notes should probably be enough to warn them. What do you think @zorun and @Glandos?

Glandos commented on ihatemoney/utils.py:
```diff
@@ -47,6 +47,31 @@ def send_email(mail_message):
+    admin email as a contact if MAIL_DEFAULT_SENDER is set to not the
+    default value and SHOW_ADMIN_EMAIL is True.
+    """
+    admin_email = current_app.config.get("MAIL_DEFAULT_SENDER")
+    error_extension = "."
+    if (
+        admin_email
+        and admin_email[1] != "admin@example.com"
+        and current_app.config.get("SHOW_ADMIN_EMAIL")
+    ):
+        error_extension = " or contact the administrator at {}.".format(admin_email[1])
```
```suggestion
    (admin_name, admin_email) = current_app.config.get("MAIL_DEFAULT_SENDER")
    error_extension = "."
    if (
        admin_email
        and admin_email != "admin@example.com"
        and current_app.config.get("SHOW_ADMIN_EMAIL")
    ):
        error_extension = f" or contact the administrator at {admin_email}."
```

f-strings are good, and we already requires Python >= 3.6
By the way, `MAIL_DEFAULT_SENDER` can't be `None`, so I destructured the assignment of it, to simplify the code.

Glandos commented on ihatemoney/utils.py:
```diff
@@ -47,6 +47,31 @@ def send_email(mail_message):
+        error_extension = " or contact the administrator at {}.".format(admin_email[1])
+
+    flash(
+        _(
+            error_message
+            + " Please check the email configuration of the server {}".format(
+                error_extension
+            )
+        ),
```
```suggestion
        _(
            f"{error_message} Please check the email configuration of the server {error_extension}"
        ),
```

Again, f-strings (may need `black` formatting though.

Glandos reviewed (changes requested):
Some suggested changes to make the code more pythonic

Glandos commented:
[**@almet**](https://github.com/almet) commented on [17 déc. 2021, 00:19 UTC+1](https://github.com/spiral-project/ihatemoney/pull/965#issuecomment-996268918 "2021-12-16T23:19:53Z - Replied by Github Reply Comments"):
> Thanks, that works for me. I'm not sure about the fact that this means that all deployed instance will get their admin email shown to the public by default without a notice to them.
> 
> I also feel that adding this in the version notes should probably be enough to warn them. What do you think [@zorun](https://github.com/zorun) and [@Glandos](https://github.com/Glandos)?

Well, to be honest, it's quite easy to get this value already: just create a dummy project, and you should get an email. If you don't have it, try to reset the password. If you still don't have it, then the value doesn't really matter, since the email configuration is probably invalid.

Glandos commented:
Thanks for your work!
